lib/pluginmanager/new.rb

```diff
@@ -0,0 +1,109 @@
+# encoding: utf-8
+require "pluginmanager/command"
+require "erb"
+require "ostruct"
+require "fileutils"
+
+class LogStash::PluginManager::New < LogStash::PluginManager::Command
+
+  parameter "TYPE",   "Type of the plugin {input, codec, filter, output}s", :attribute_name => :type
+  parameter "PLUGIN", "Name of the new plugin", :attribute_name => :plugin_name
+
+  TYPES = [ "input", "filter", "output" ]
+
+  def execute
+    validate_parameters
+
+    source = File.join(File.dirname(__FILE__), "templates", "#{type}-plugin")
+    target = FileUtils.mkdir(File.join(Dir.pwd, full_plugin_name))
+    puts " #{ green("Creating") } #{full_plugin_name}"
+
+    create_scaffold(source, target)
+  end
+
+  private
+
+  def create_scaffold(source, target)
+    render_options(options)
+    transform_r(source, target)
+  end
+
+  def transform_r(source, target)
+    Dir.entries(source).each do |entry|
+      next if [ ".", ".." ].include?(entry)
+      source_entry = File.join(source, entry)
+      target_entry = File.join(target, entry)
+
+      if File.directory?(source_entry)
+        FileUtils.mkdir(target_entry) unless File.exists?(target_entry)
+        transform_r(source_entry, target_entry)
+      else
+        # copy the new file, in case of being an .erb file should render first
+        if source_entry.end_with?("erb")
+          target_entry = target_entry.gsub(/.erb$/,"").gsub("example", plugin_name)
+          File.open(target_entry, "w") { |f| f.write(render(source_entry)) }
+        else
+          FileUtils.cp(source_entry, target_entry)
+        end
+        puts "\t #{ green("create") } #{File.join(full_plugin_name, File.basename(target_entry))}"
+      end
+    end
+  end
+
+  def classify(plugin_name)
+    plugin_name.split(/-|_/).map { |e| e.capitalize }.join("")
+  end
+
+  def render_options(options)
+    options.each_pair do |name, value|
+      instance_variable_set("@#{name}", value)
+    end
+  end
+
+  def render(source)
+    template = File.read(source)
+    renderer = ERB.new(template)
+    renderer.result(binding())
```

Comment:
There is a small problem here by using instance variables, if the variable doesn't exist it will return a `NIL` values which is not really what we want. I beleive when a non existing value is used in a template the rendering code should fail violently. This has the benefit of the adding unnecessary variables to the `new` command class.

I suggest we pass a `RenderContext` object to the erb rendered and replace the instance variables with method calls.

```ruby
require "erb"
class RenderContext
   def initialize(options = {})
     options.each do |key, value|
       define_singleton_method(key) { value }
     end
   end

   def get_binding
     binding()
   end
end

erb = ERB.new("Hello <%= name%>")
puts erb.result(RenderContext.new({ :name => "PH" }).get_binding)

erb = ERB.new("Hello <%= patate %>")
puts erb.result(RenderContext.new({ :name => "PH" }).get_binding)
```

Result:
```
Hello PH
NameError: undefined local variable or method `patate' for #<RenderContext:0x6b143ee9>
  result at (erb):1
    eval at org/jruby/RubyKernel.java:1107
  result at /Users/ph/.rvm/rubies/jruby-1.7.19/lib/ruby/1.9/erb.rb:838
  (root) at es/experiments/erb.rb:19
```